File: py-server/app/controllers/auth_controller.py
import os
import bcrypt
from flask import request, jsonify
from app.models.user import User
from app.helpers.rekognition_helper import compare_faces


def authenticate():
    try:
        data = request.get_json()
        username_or_email = data.get('username_or_email')
        password = data.get('password')
        
        # Validar la entrada
        if not username_or_email or not password:
            return jsonify({'message': 'Todos los campos son necesarios.'}), 400
        
        # Buscar el usuario primero por username
        user = User.query.filter_by(username=username_or_email).first()
        
        # Si no se encontró por username, intentar buscar por email
        if not user:
            user = User.query.filter_by(email=username_or_email).first()
        
        if not user:
            return jsonify({'message': 'Usuario no encontrado.'}), 404
        
        # Comparar la contraseña
        if not bcrypt.checkpw(password.encode('utf-8'), user.password.encode('utf-8')):
            return jsonify({'message': 'La contraseña es incorrecta.'}), 401
        
        return jsonify(user.to_dict()), 200   
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500


def authenticate_with_facial_recognition():
    try:
        username_or_email = request.form.get('username_or_email')
        image = request.files.get('image')

        # Validar la entrada
        if not username_or_email or not image:
            return jsonify({'message': 'Todos los campos son necesarios.'}), 400

        # Buscar el usuario primero por username
        user = User.query.filter_by(username=username_or_email).first()
        
        # Si no se encontró por username, intentar buscar por email
        if not user:
            user = User.query.filter_by(email=username_or_email).first()
        
        if not user:
            return jsonify({'message': 'Usuario no encontrado.'}), 404

        # Verificar si el usuario tiene activado el reconocimiento facial
        if not user.login_image:
            return jsonify({'message': 'El reconocimiento facial está desactivado para este usuario.'}), 403

        # Obtener el nombre del bucket de la variable de entorno
        s3_bucket_name = os.getenv('AWS_BUCKET_NAME')

        # Extraer el nombre del objeto S3 desde la URL de image_key del usuario
        s3_object_name = user.image_key.split('/', 3)[-1]

        # Comparar imágenes
        if not compare_faces(image, s3_bucket_name, s3_object_name):
            return jsonify({'message': 'Los rostros no coinciden.'}), 404

        return jsonify(user.to_dict()), 200
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500

File: js-server/controllers/auth_controller.js
const express = require('express');
const bcrypt = require('bcrypt');
const multer = require('multer');
const AWS = require('aws-sdk');
const dotenv = require('dotenv');
const conexion  = require('../conexion/conexion')


const authenticate = (req, res) => {
    const { username_or_email, password } = req.body;

    if (!username_or_email || !password) {
        return res.status(400).json({ message: 'Todos los campos son necesarios.' });
    }

    // Buscar el usuario por username o email
    conexion.query('SELECT * FROM Usuario WHERE nombre_usuario = ? OR correo = ?', [username_or_email, username_or_email], (error, results) => {
        if (error) {
            console.error('Error al buscar el usuario:', error);
            return res.status(500).json({ message: `Error interno del servidor: ${error.message}` });
        }

        const user = results[0];

        if (!user) {
            return res.status(404).json({ message: 'Usuario no encontrado.' });
        }

        // Comparar la contraseña
        bcrypt.compare(password, user.pass, (error, isMatch) => {
            if (error) {
                console.error('Error al comparar las contraseñas:', error);
                return res.status(500).json({ message: `Error interno del servidor: ${error.message}` });
            }

            if (!isMatch) {
                return res.status(401).json({ message: 'La contraseña es incorrecta.' });
            }

            // Devolver la información del usuario
            const userToDict = {
                user_id: user.id_usuario,
                username: user.nombre_usuario,
                email: user.correo,
                image: user.imagen,
                login_image: user.login_imagen,
                image_key: user.imagen_clave
            };

            return res.status(200).json(userToDict);
        });
    });
};


module.exports = { authenticate };